src/ucs/datastruct/array.h

```diff
@@ -0,0 +1,224 @@
+ */
+#define UCS_ARRAY_FIXED_INITIALIZER(_buffer, _capacity) \
+    { (_buffer), 0, \
+      ((_capacity) & UCS_ARRAY_CAP_MASK) | UCS_ARRAY_CAP_FLAG_FIXED }
+
+
+/**
+ * Expands to array type definition
+ *
+ * @param _name Array name (as passed to UCS_ARRAY_DECLARE)
+ *
+ * Example:
+ *
+ * @code{.c}
+ * ucs_array_t(int_array) my_array;
+ * @endcode
+ */
+#define ucs_array_t(_name) \
+    UCS_ARRAY_IDENTIFIER(_name, _t)
+
+
+/**
+ * Initialize a dynamic array. Such array can grow automatically to accommodate
+ * for more elements.
+ *
+ * @param _name    Array name
+ * @param _array   Pointer to the array to initialize
+ */
+#define ucs_array_init_dynamic(_name, _array) \
+    UCS_ARRAY_IDENTIFIER(_name, _init_dynamic)(_array)
+
+
+/*
+ * Cleanup a dynamic array.
+ *
+ * @param _name    Array name
+ * @param _array   Array to cleanup
+ */
+#define ucs_array_cleanup_dynamic(_name, _array) \
+    UCS_ARRAY_IDENTIFIER(_name, _cleanup_dynamic)(_array)
+
+
+/*
+ * Grow the array memory buffer to the at least required capacity. This does not
+ * change the array length or existing contents, but the backing buffer can be
+ * relocated to another memory area.
+ *
+ * @param _name    Array name
+ * @param _array   Array to reserve buffer for
+ *
+ * @return UCS_OK if successful, UCS_ERR_NO_MEMORY if cannot grow the array
+ */
+#define ucs_array_reserve(_name, _array, _min_capacity) \
+    UCS_ARRAY_IDENTIFIER(_name, _reserve)(_array, _min_capacity)
+
+
+/*
+ * Add an element to the end of the array and return its index.
+ *
+ * @param _name     Array name
+ * @param _array    Array to add element to
+ * @param _index_p  Filled with the index of the added element
+ *
+ * @return UCS_OK if added, UCS_ERR_NO_MEMORY if cannot grow the array
+ */
+#define ucs_array_append(_name, _array, _index_p) \
+   UCS_ARRAY_IDENTIFIER(_name, _append)(_array, _index_p)
+
+
+/**
+ * @return Number of elements in the array
+ */
+#define ucs_array_length(_array) \
+    ((_array)->length)
+
+
+/**
+ * @return Current capacity of the array
+ */
+#define ucs_array_capacity(_array) \
+    ((_array)->capacity & UCS_ARRAY_CAP_MASK)
+
+
+/**
+ * @return Whether this is a fixed-length array
+ */
+#define ucs_array_is_fixed(_array) \
+    ((_array)->capacity & UCS_ARRAY_CAP_FLAG_FIXED)
+
+
+/**
+ * Get array element
+ *
+ * @param _array   Array whose element to retrieve
+ * @param _index   Element index in the array
+ *
+ * @return L-value of a specified element in the array
+ */
+#define ucs_array_elem(_array, _index) \
+    ((_array)->buffer[_index])
+
+
+/**
+ * @return Whether the array is empty
+ */
+#define ucs_array_is_empty(_array) \
+    (ucs_array_length(_array) == 0)
+
+
+/**
+ * @return Pointer to array end element (first non-valid element)
+ */
+#define ucs_array_end(_array) \
+    ((_array)->buffer + ucs_array_length(_array))
+
+
+/**
+ * @return How many elements could be added within the current array capacity
+ */
+#define ucs_array_end_capacity(_array) \
```

Comment:
maybe `ucs_array_current_capacity`?

Reply:
it does not return returns the whole capacity, but only the "remainder"
("current" capacity can be confused with overall capacity)

Reply:
then maybe `ucs_array_remain_capacity` or something? Imo, `ucs_array_end_capacity` sounds confusing either

Reply:
how about ucs_array_available_size?

Reply:
sounds good